dir.c

```diff
@@ -758,13 +758,33 @@ static int add_excludes(const char *fname, const char *base, int baselen,
 	size_t size = 0;
 	char *buf, *entry;
 
-	if (is_fscache_enabled()) {
+	if (is_fscache_enabled(fname)) {
+		/*
+		 * A performance optimization for status.
+		 *
+		 * When fscache is enabled and we have a non-global path, the
+		 * fscache database should already know if a .gitignore file
+		 * exists in the worktree.  We let the fscache-lstat helper
+		 * do the lookup *before* we try to open it.  Since very few
+		 * directories contain .gitignore files, we can speed up status
+		 * because we don't have to touch the filesystem here.
+		 *
+		 * We cannot use the fscache-stat helper because it needs to
+		 * actually open the file to dereference symlinks and then do
+		 * fstat.  This defeats our purpose here.
+		 *
+		 * If the .gitignore file is successfully opened and in the
+		 * rare case that it is a symlink, we fstat it to replace the
+		 * original lstat stat-data to complete our illusion.
+		 */
 		if (lstat(fname, &st) < 0) {
```

Comment:
I wonder if all that is needed is an `s/lstat/stat/`?

Reply:
no. see updated commit message.